### PR TITLE
Update way, closestWay with restrictions

### PR DESCRIPTION
**Pull request by Royar13**

### CODE REVIEW

File: GIS.cpp [hunk 1]
```diff
@@ -127,50 +110,42 @@ std::pair<Coordinates, EntityId> GIS::getWayClosestPoint(const Coordinates &coor
...
-                {
-                    wayFound = true;
-                    Coordinates candidate = entity.getGeometry()->getClosestPoint(coord);
-                    Meters distance = CoordinatesMath::calculateDistance(candidate, coord);
-                    if (distance < shortestDistance)
-                    {
-                        closest = candidate;
-                        shortestDistance = distance;
-                        closestEntityId = entityId;
+                Entity *entity = entities.find(entityId)->second.get();
```
Comment (talsch64) [inline GIS.cpp:124]: better not to use get

File: entities/Entity.h [hunk 2]
```diff
@@ -48,7 +46,7 @@ class Entity {
 
     void setId(const EntityId &id);
 
-    const std::unique_ptr<Geometry> &getGeometry() const;
+    virtual const Geometry *getGeometry() const = 0;
```
Comment (talsch64) [inline entities/Entity.h:49]: Better to use shared_ptr then raw pointer.

File: entities/Way.h [hunk 3]
```diff
@@ -4,22 +4,25 @@
...
-class Way: public Entity {
+class Way : public Entity {
     EntityId from;
     EntityId to;
+    std::unique_ptr<PointList> geometry;
     std::string direction;
     int speedLimit;
     bool tollRoad;
     std::vector<std::string> restricted;
+    std::optional<Meters> length;
```
Comment (talsch64) [inline entities/Way.h:21]: why optional?

**Review by talsch64 — CHANGES_REQUESTED** [review]

File: GIS.cpp (hunk 1 above)
Comment (Royar13) [reply GIS.cpp]: Fixed

File: entities/Entity.h (hunk 2 above)
Comment (Royar13) [reply entities/Entity.h]: Fixed

**Pull request merged**